### PR TITLE
Asst al 3

### PR DESCRIPTION
**Pull request by russellschmidt**

### CODE REVIEW

File: 02-algorithms/03-sorting/bucket-sort.rb [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+			bucketGHI << element
+		elsif element < "M"
+			bucketJKL << element
+		elsif element < "P"
+			bucketMNO << element
+		elsif element < "S"
+			bucketPQR << element
+		elsif element < "V"
+			bucketSTU << element
+		elsif element <= "Zzzzzz"
```
Comment (dawsonc623) [inline 02-algorithms/03-sorting/bucket-sort.rb:29]: What if I wanted to put in "Zzzzzzzzzzzz"!?

File: 02-algorithms/03-sorting/bucket-sort.rb [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+			bucketVWXYZ << element
+		end
+	end
+
+	# sort within buckets
+	megaBucket << bucketABC << bucketDEF << bucketGHI << bucketJKL << bucketMNO << bucketPQR << bucketSTU << bucketVWXYZ
```
Comment (dawsonc623) [inline 02-algorithms/03-sorting/bucket-sort.rb:35]: New! From KFC!

File: 02-algorithms/03-sorting/heap-sort.rb [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+# heapsort - pop off the root each time and put into an array
+def heap_sort(collection)
+  sorted_list = []
+
+  root_index = find_root_index(collection)
+  root = Node.new(collection.delete_at(root_index))
+  heap = min_heapify(collection, root)
+
+  heap.lowest_value(nil, sorted_list)
```
Comment (dawsonc623) [inline 02-algorithms/03-sorting/heap-sort.rb:36]: All you have to do is add the root to the sorted list and then delete it. The delete operation will ensure the next lowest is at the top. Then keep doing that.

File: 02-algorithms/03-sorting/bucket-sort.rb (hunk 1 above)
Comment (russellschmidt) [reply 02-algorithms/03-sorting/bucket-sort.rb:29]: Fair point. Should be 'else' for the last bucket, then.

File: 02-algorithms/03-sorting/heap-sort.rb (hunk 3 above)
Comment (russellschmidt) [reply 02-algorithms/03-sorting/heap-sort.rb:36]: I didn't know if it was bad form to delete the heap I made. Such a nice heap.

File: 02-algorithms/03-sorting/bucket-sort.rb (hunk 1 above)
Comment (dawsonc623) [reply 02-algorithms/03-sorting/bucket-sort.rb:29]: Nailed it.

File: 02-algorithms/03-sorting/heap-sort.rb (hunk 3 above)
Comment (dawsonc623) [reply 02-algorithms/03-sorting/heap-sort.rb:36]: It's entire purpose, and the entire purpose of the heap data structure, is to be slowly ripped apart, root by root, until a linearly sorted alternative data structure is left. Kind of tragic, really.

File: 02-algorithms/03-sorting/answers.txt [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+.020		X			X				X				X					X						X					x
+.010		X			X				X				X					X						X					X
+------------------------------------------------------------------
+			heap	quick		bucket	merge		insertion		selection		bubble
+
+
+QUESTION 5
+Based on the chart and the results, I think that overall, you can see my implementation of various sort algorithms is likely the culprit as to why certain sort algorithms are faster than others - the three I created the code for are notably slower than the ones where Bloc provided the code.
+
+However, in general, the sorting algorithms that are iterating just a couple of times over an array in a single function call (insertion, selection, bubble) might be faster for small data sets where the speed of array operations operating within a single contiguous block of memory enjoy an advantage over algorithms where you have multiple function calls per iteration and/or multiple arrays being referenced. Thus the increased complexity of these multi-array and/or multi-function implementations makes them relatively slower for these smaller data sets.
```
Comment (dawsonc623) [inline 02-algorithms/03-sorting/answers.txt:76]: That is a good (and valid) point. Some algorithms have enough overhead from set up and the like that even if each iteration is faster, there is a certain number of iterations that need to pass before it is worthwhile. I will look back to the 1000000 item search analysis I gave on handling unsorted collections.

**Comment by dawsonc623** [comment]
Branch is safe to delete.

**Pull request merged**